File: src/Python/src/vladimir_reshetnikov/data_structures/finger_tree/persistent_chunked_bit_set.py
"""Sparse persistent bit set over the shared measured sequence."""

from __future__ import annotations

from collections.abc import Iterable, Iterator
from dataclasses import dataclass
from typing import Literal

from .measured_sequence import MeasuredSequence, SequenceLocate
from .measures import OptionalValue

_MAXIMUM_BIT_INDEX = (1 << 31) - 1
_WORD_MASK = (1 << 64) - 1


@dataclass(frozen=True, slots=True)
class _BitSetChunk:
    word_index: int
    bits: int


@dataclass(frozen=True, slots=True)
class _BitSetSummary:
    chunk_count: int
    pop_count: int
    last_word_index: OptionalValue[int]


class _BitSetMeasure:
    identity = _BitSetSummary(0, 0, OptionalValue.none())

    def combine(self, left: _BitSetSummary, right: _BitSetSummary) -> _BitSetSummary:
        return _BitSetSummary(
            left.chunk_count + right.chunk_count,
            left.pop_count + right.pop_count,
            right.last_word_index if right.last_word_index.has_value else left.last_word_index,
        )

    def measure(self, element: _BitSetChunk) -> _BitSetSummary:
        return _BitSetSummary(1, element.bits.bit_count(), OptionalValue.some(element.word_index))


_BIT_SET_MEASURE = _BitSetMeasure()
_Operation = Literal["union", "intersect", "except", "symmetric_except"]


@dataclass(frozen=True, slots=True)
class ChunkedBitSetUpdateResult:
    """Nonthrowing point-update result."""

    changed: bool
    set: PersistentChunkedBitSet


@dataclass(frozen=True, slots=True)
class ChunkedBitSetStatistics:
    """Validated sparse-word statistics."""

    chunk_count: int
    pop_count: int


class PersistentChunkedBitSet:
    """Immutable sparse bit set over the nonnegative signed-32-bit domain.

    Only nonzero 64-bit words are represented. Cached population and last-word summaries provide
    logarithmic point edits, inclusive rank, and zero-based select in represented words.
    """

    __slots__ = ("_chunks",)

    def __init__(self, chunks: MeasuredSequence[_BitSetChunk, _BitSetSummary]) -> None:
        self._chunks = chunks

    @classmethod
    def empty(cls) -> PersistentChunkedBitSet:
        return _EMPTY_BIT_SET

    @classmethod
    def from_values(cls, bit_indexes: Iterable[int]) -> PersistentChunkedBitSet:
        if bit_indexes is None:
            raise TypeError("bit_indexes must be iterable.")
        words: dict[int, int] = {}
        for bit_index in bit_indexes:
            _require_bit_index(bit_index)
            word_index = bit_index >> 6
            words[word_index] = words.get(word_index, 0) | (1 << (bit_index & 63))
        if not words:
            return cls.empty()
        chunks = [_BitSetChunk(index, words[index]) for index in sorted(words)]
        return cls(MeasuredSequence.from_iterable(chunks, _BIT_SET_MEASURE))

    @property
    def count(self) -> int:
        return self._chunks.measure.pop_count

    @property
    def chunk_count(self) -> int:
        return self._chunks.measure.chunk_count

    @property
    def is_empty(self) -> bool:
        return self._chunks.is_empty

    def __len__(self) -> int:
        return self.count

    def __bool__(self) -> bool:
        return not self.is_empty

    def contains(self, bit_index: int) -> bool:
        if type(bit_index) is not int or bit_index < 0 or bit_index > _MAXIMUM_BIT_INDEX:
            return False
        word_index = bit_index >> 6
        located = self._locate_word(word_index)
        return bool(
            located.found
            and located.value is not None
            and located.value.word_index == word_index
            and located.value.bits & (1 << (bit_index & 63))
        )

    def __contains__(self, bit_index: object) -> bool:
        return self.contains(bit_index) if type(bit_index) is int else False

    def add(self, bit_index: int) -> PersistentChunkedBitSet:
        _require_bit_index(bit_index)
        word_index = bit_index >> 6
        bit = 1 << (bit_index & 63)
        located = self._locate_word(word_index)
        if located.found and located.value is not None and located.value.word_index == word_index:
            bits = located.value.bits | bit
            if bits == located.value.bits:
                return self
            chunks = self._chunks.set_at(located.index, _BitSetChunk(word_index, bits))
        else:
            chunks = self._chunks.insert_at(located.index, _BitSetChunk(word_index, bit))
        if chunks is None:
            raise AssertionError("A validated bit-set edit failed.")
        return PersistentChunkedBitSet(chunks)

    def try_add(self, bit_index: int) -> ChunkedBitSetUpdateResult:
        result = self.add(bit_index)
        return ChunkedBitSetUpdateResult(result is not self, result)

    def remove(self, bit_index: int) -> PersistentChunkedBitSet:
        if type(bit_index) is not int or bit_index < 0 or bit_index > _MAXIMUM_BIT_INDEX:
            return self
        word_index = bit_index >> 6
        bit = 1 << (bit_index & 63)
        located = self._locate_word(word_index)
        if (
            not located.found
            or located.value is None
            or located.value.word_index != word_index
            or not located.value.bits & bit
        ):
            return self
        bits = located.value.bits & ~bit
        chunks = (
            self._chunks.remove_at(located.index)
            if bits == 0
            else self._chunks.set_at(located.index, _BitSetChunk(word_index, bits))
        )
        if chunks is None:
            raise AssertionError("A validated bit-set removal failed.")
        return PersistentChunkedBitSet(chunks)

    def try_remove(self, bit_index: int) -> ChunkedBitSetUpdateResult:
        result = self.remove(bit_index)
        return ChunkedBitSetUpdateResult(result is not self, result)

    def rank(self, bit_index: int) -> int:
        """Count set indexes less than or equal to *bit_index*."""

        if type(bit_index) is not int:
            raise TypeError("bit_index must be an integer.")
        if bit_index < 0:
            return 0
        if bit_index >= _MAXIMUM_BIT_INDEX:
            return self.count
        word_index = bit_index >> 6
        located = self._locate_word(word_index)
        if not located.found or located.value is None or located.value.word_index != word_index:
            return located.measure_before.pop_count
        offset = bit_index & 63
        mask = _WORD_MASK if offset == 63 else (1 << (offset + 1)) - 1
        return located.measure_before.pop_count + (located.value.bits & mask).bit_count()

    def select(self, rank: int) -> int:
        selected = self.try_select(rank)
        if selected is None:
            raise IndexError("rank must identify a set bit.")
        return selected

    def try_select(self, rank: int) -> int | None:
        if type(rank) is not int or rank < 0 or rank >= self.count:
            return None
        located = self._chunks.locate(lambda summary: summary.pop_count > rank)
        if not located.found or located.value is None:
            return None
        bits = located.value.bits
        for _ in range(rank - located.measure_before.pop_count):
            bits &= bits - 1
        offset = (bits & -bits).bit_length() - 1
        return (located.value.word_index << 6) + offset

    def union(self, other: PersistentChunkedBitSet) -> PersistentChunkedBitSet:
        return self if other is self or other.is_empty else self._combine(other, "union")

    def intersect(self, other: PersistentChunkedBitSet) -> PersistentChunkedBitSet:
        return self if other is self else self._combine(other, "intersect")

    def except_(self, other: PersistentChunkedBitSet) -> PersistentChunkedBitSet:
        return self.empty() if other is self else self._combine(other, "except")

    def symmetric_except(self, other: PersistentChunkedBitSet) -> PersistentChunkedBitSet:
        return self.empty() if other is self else self._combine(other, "symmetric_except")

    def clear(self) -> PersistentChunkedBitSet:
        return self if self.is_empty else self.empty()

    def __iter__(self) -> Iterator[int]:
        for chunk in self._chunks:
            bits = chunk.bits
            while bits:
                lowest = bits & -bits
                yield (chunk.word_index << 6) + lowest.bit_length() - 1
                bits &= bits - 1

    def shares_storage_with(self, other: PersistentChunkedBitSet) -> bool:
        return self._chunks.shares_structure_with(other._chunks)

    def validate_structure(self) -> ChunkedBitSetStatistics:
        chunk_count = 0
        pop_count = 0
        previous = -1
        for chunk in self._chunks:
            if chunk.word_index <= previous or chunk.bits <= 0 or chunk.bits > _WORD_MASK:
                raise ValueError("The chunked bit set has invalid word order or contents.")
            previous = chunk.word_index
            chunk_count += 1
            pop_count += chunk.bits.bit_count()
        summary = self._chunks.measure
        expected_last = OptionalValue.none() if chunk_count == 0 else OptionalValue.some(previous)
        if (
            summary.chunk_count != chunk_count
            or summary.pop_count != pop_count
            or summary.last_word_index != expected_last
        ):
            raise ValueError("The chunked bit-set annotations disagree with its words.")
        return ChunkedBitSetStatistics(chunk_count, pop_count)

    def _locate_word(self, word_index: int) -> SequenceLocate[_BitSetChunk, _BitSetSummary]:
        return self._chunks.locate(
            lambda summary: (
                summary.last_word_index.has_value and summary.last_word_index.unwrap() >= word_index
            )
        )

    def _combine(
        self, other: PersistentChunkedBitSet, operation: _Operation
    ) -> PersistentChunkedBitSet:
        left = self._chunks.to_list()
        right = other._chunks.to_list()
        result: list[_BitSetChunk] = []
        left_index = 0
        right_index = 0
        while left_index < len(left) or right_index < len(right):
            left_chunk = left[left_index] if left_index < len(left) else None
            right_chunk = right[right_index] if right_index < len(right) else None
            if right_chunk is None or (
                left_chunk is not None and left_chunk.word_index < right_chunk.word_index
            ):
                if operation in ("union", "except", "symmetric_except"):
                    if left_chunk is None:
                        raise AssertionError("The left word stream ended unexpectedly.")
                    result.append(left_chunk)
                left_index += 1
                continue
            if left_chunk is None or right_chunk.word_index < left_chunk.word_index:
                if operation in ("union", "symmetric_except"):
                    result.append(right_chunk)
                right_index += 1
                continue
            bits = (
                left_chunk.bits | right_chunk.bits
                if operation == "union"
                else left_chunk.bits & right_chunk.bits
                if operation == "intersect"
                else left_chunk.bits & ~right_chunk.bits
                if operation == "except"
                else left_chunk.bits ^ right_chunk.bits
            )
            if bits:
                result.append(_BitSetChunk(left_chunk.word_index, bits))
            left_index += 1
            right_index += 1
        if result == left:
            return self
        if not result:
            return self.empty()
        return PersistentChunkedBitSet(MeasuredSequence.from_iterable(result, _BIT_SET_MEASURE))


def _require_bit_index(bit_index: int) -> None:
    if type(bit_index) is not int or bit_index < 0 or bit_index > _MAXIMUM_BIT_INDEX:
        raise ValueError("bit_index must be a nonnegative signed-32-bit integer.")


_EMPTY_BIT_SET = PersistentChunkedBitSet(MeasuredSequence.empty(_BIT_SET_MEASURE))

__all__ = [
    "ChunkedBitSetStatistics",
    "ChunkedBitSetUpdateResult",
    "PersistentChunkedBitSet",
]

File: src/TypeScript/src/finger-tree/persistent-chunked-bit-set.ts
/**
 * Persistent sparse bit set stored as measured fixed-width chunks.
 *
 * Only chunks containing a set bit are stored, and the measure caches each subtree's population
 * count, which is what makes point lookup, inclusive rank, and select logarithmic rather than
 * linear in the highest index ever set. Indices are confined to the cross-language nonnegative
 * signed 32-bit domain.
 */
import type { MeasurePolicy } from "./measures.js";
import { MeasuredSequence } from "./measured-sequence.js";

interface BitSetChunk {
    readonly wordIndex: number;
    readonly bits: bigint;
}

interface BitSetSummary {
    readonly chunkCount: number;
    readonly popCount: number;
    readonly lastWordIndex: number | undefined;
}

const emptySummary: BitSetSummary = { chunkCount: 0, popCount: 0, lastWordIndex: undefined };

const bitSetMeasure: MeasurePolicy<BitSetChunk, BitSetSummary> = {
    identity: emptySummary,
    combine: (left, right): BitSetSummary => ({
        chunkCount: left.chunkCount + right.chunkCount,
        popCount: left.popCount + right.popCount,
        lastWordIndex: right.lastWordIndex ?? left.lastWordIndex,
    }),
    measure: (chunk): BitSetSummary => ({
        chunkCount: 1,
        popCount: popCount(chunk.bits),
        lastWordIndex: chunk.wordIndex,
    }),
};

const maximumBitIndex = 0x7fff_ffff;

type BitSetOperation = "union" | "intersect" | "except" | "symmetricExcept";

/** Result of a nonthrowing chunked-bit-set point update. */
export interface ChunkedBitSetUpdateResult {
    /** Whether the bit's state actually changed. */
    readonly changed: boolean;
    /** A map with the key bound to the value, adding or replacing as needed. */
    readonly set: PersistentChunkedBitSet;
}

/** Structural statistics for the sparse measured representation. */
export interface ChunkedBitSetStatistics {
    /** Number of stored chunks, a measure of representation size rather than of contents. */
    readonly chunkCount: number;
    /** Number of set bits, recounted from the stored chunks. */
    readonly popCount: number;
}

/** The outcome of seeking a cursor to a bit index; on a miss it sits before the next set bit. */
export interface ChunkedBitSetCursorSearch {
    /** Whether the probed index is set. */
    readonly found: boolean;
    /** A cursor at the given gap of the collection. */
    readonly cursor: PersistentChunkedBitSetCursor;
}

/**
 * Immutable sparse bit set over the nonnegative signed-32-bit index domain.
 *
 * Only nonzero 64-bit words are represented. Cached ordered-word and population summaries support
 * logarithmic membership, point edits, inclusive rank, and zero-based select in represented words.
 */
export class PersistentChunkedBitSet implements Iterable<number> {
    static readonly #empty = new PersistentChunkedBitSet(MeasuredSequence.empty(bitSetMeasure));
    readonly #chunks: MeasuredSequence<BitSetChunk, BitSetSummary>;

    private constructor(chunks: MeasuredSequence<BitSetChunk, BitSetSummary>) { this.#chunks = chunks; }

    /** The empty bit set, retaining the supplied policy objects. */
    public static empty(): PersistentChunkedBitSet { return PersistentChunkedBitSet.#empty; }

    /** Build a bit set from the given set bits. */
    public static from(bitIndexes: Iterable<number>): PersistentChunkedBitSet {
        if (bitIndexes === null || bitIndexes === undefined) throw new TypeError("bitIndexes must be iterable.");
        const words = new Map<number, bigint>();
        for (const bitIndex of bitIndexes) {
            requireBitIndex(bitIndex);
            const wordIndex = Math.floor(bitIndex / 64);
            const bit = 1n << BigInt(bitIndex & 63);
            words.set(wordIndex, (words.get(wordIndex) ?? 0n) | bit);
        }
        if (words.size === 0) return PersistentChunkedBitSet.empty();
        const chunks = Array.from(words, ([wordIndex, bits]): BitSetChunk => ({ wordIndex, bits }))
            .sort((left, right) => left.wordIndex - right.wordIndex);
        return new PersistentChunkedBitSet(MeasuredSequence.from(chunks, bitSetMeasure));
    }

    /** Number of set bits. */
    public get count(): number { return this.#chunks.measure.popCount; }
    /** Number of set bits. */
    public get size(): number { return this.count; }
    /** Number of stored chunks, a measure of representation size rather than of contents. */
    public get chunkCount(): number { return this.#chunks.measure.chunkCount; }
    /** Whether the bit set holds no set bits. */
    public get isEmpty(): boolean { return this.#chunks.isEmpty; }

    /** Whether the set bit is present. */
    public contains(bitIndex: number): boolean {
        if (!Number.isInteger(bitIndex) || bitIndex < 0 || bitIndex > maximumBitIndex) return false;
        const wordIndex = Math.floor(bitIndex / 64);
        const located = this.locateWord(wordIndex);
        return located.found && located.value?.wordIndex === wordIndex
            && (located.value.bits & (1n << BigInt(bitIndex & 63))) !== 0n;
    }

    /** A bit set containing the given set bit; returns the receiver when it is already present. */
    public add(bitIndex: number): PersistentChunkedBitSet {
        requireBitIndex(bitIndex);
        const wordIndex = Math.floor(bitIndex / 64);
        const bit = 1n << BigInt(bitIndex & 63);
        const located = this.locateWord(wordIndex);
        if (located.found && located.value?.wordIndex === wordIndex) {
            const bits = located.value.bits | bit;
            return bits === located.value.bits
                ? this
                : new PersistentChunkedBitSet(this.#chunks.setAt(located.index, { wordIndex, bits })!);
        }
        return new PersistentChunkedBitSet(this.#chunks.insertAt(located.index, { wordIndex, bits: bit })!);
    }

    /** Add the entry, reporting whether it was added rather than throwing on a duplicate. */
    public tryAdd(bitIndex: number): ChunkedBitSetUpdateResult {
        const set = this.add(bitIndex);
        return { changed: set !== this, set };
    }

    /** A bit set without that set bit; returns the receiver when it is absent. */
    public remove(bitIndex: number): PersistentChunkedBitSet {
        if (!Number.isInteger(bitIndex) || bitIndex < 0 || bitIndex > maximumBitIndex) return this;
        const wordIndex = Math.floor(bitIndex / 64);
        const bit = 1n << BigInt(bitIndex & 63);
        const located = this.locateWord(wordIndex);
        if (!located.found || located.value?.wordIndex !== wordIndex || (located.value.bits & bit) === 0n) return this;
        const bits = located.value.bits & ~bit;
        return new PersistentChunkedBitSet(
            bits === 0n
                ? this.#chunks.removeAt(located.index)!
                : this.#chunks.setAt(located.index, { wordIndex, bits })!,
        );
    }

    /** Remove the set bit and report what was removed, or `undefined` when absent. */
    public tryRemove(bitIndex: number): ChunkedBitSetUpdateResult {
        const set = this.remove(bitIndex);
        return { changed: set !== this, set };
    }

    /** Counts set indexes less than or equal to `bitIndex`. */
    public rank(bitIndex: number): number {
        if (!Number.isFinite(bitIndex) || bitIndex < 0) return 0;
        if (bitIndex >= maximumBitIndex) return this.count;
        const integral = Math.floor(bitIndex);
        const wordIndex = Math.floor(integral / 64);
        const located = this.locateWord(wordIndex);
        if (!located.found || located.value?.wordIndex !== wordIndex) return located.measureBefore.popCount;
        const offset = integral & 63;
        const mask = offset === 63 ? (1n << 64n) - 1n : (1n << BigInt(offset + 1)) - 1n;
        return located.measureBefore.popCount + popCount(located.value.bits & mask);
    }

    /** Returns the bit at a zero-based population rank. */
    public select(rank: number): number {
        const selected = this.trySelect(rank);
        if (selected === undefined) throw new RangeError("rank must be an integer from zero through count minus one.");
        return selected;
    }

    /** The bit index at the given zero-based population rank, or `undefined` when out of range. */
    public trySelect(rank: number): number | undefined {
        if (!Number.isInteger(rank) || rank < 0 || rank >= this.count) return undefined;
        const located = this.#chunks.locate((summary) => summary.popCount > rank);
        if (!located.found || located.value === undefined) return undefined;
        let bits = located.value.bits;
        const localRank = rank - located.measureBefore.popCount;
        for (let index = 0; index < localRank; index++) bits &= bits - 1n;
        let offset = 0;
        while ((bits & 1n) === 0n) { bits >>= 1n; offset++; }
        return located.value.wordIndex * 64 + offset;
    }

    /** The set bits of both bit sets. */
    public union(other: PersistentChunkedBitSet): PersistentChunkedBitSet {
        return other === this || other.isEmpty ? this : this.combine(other, "union");
    }

    /** The set bits present in both bit sets. */
    public intersect(other: PersistentChunkedBitSet): PersistentChunkedBitSet {
        return other === this ? this : this.combine(other, "intersect");
    }

    /** This bit set's set bits that are absent from the other. */
    public except(other: PersistentChunkedBitSet): PersistentChunkedBitSet {
        return other === this ? PersistentChunkedBitSet.empty() : this.combine(other, "except");
    }

    /** The set bits present in exactly one of the two bit sets. */
    public symmetricExcept(other: PersistentChunkedBitSet): PersistentChunkedBitSet {
        return other === this ? PersistentChunkedBitSet.empty() : this.combine(other, "symmetricExcept");
    }

    /** An empty bit set retaining the same policies; returns the receiver when already empty. */
    public clear(): PersistentChunkedBitSet { return this.isEmpty ? this : PersistentChunkedBitSet.empty(); }

    /** A cursor at the given gap of the bit set. */
    public cursorAt(position = 0): PersistentChunkedBitSetCursor {
        return new PersistentChunkedBitSetCursor(this, position);
    }

    /** A cursor before the first set bit at or after the given index. */
    public cursorAtOrAfter(bitIndex: number): PersistentChunkedBitSetCursor {
        const position = bitIndex <= 0 ? 0 : this.rank(bitIndex - 1);
        return this.cursorAt(position);
    }

    /**
     * Seek to the probe and report whether it is present. On a miss the cursor sits where the probe
     * would be inserted and remains usable.
     */
    public findCursor(bitIndex: number): ChunkedBitSetCursorSearch {
        const cursor = this.cursorAtOrAfter(bitIndex);
        return { found: bitIndex >= 0 && cursor.peekNext()?.value === bitIndex, cursor };
    }

    public *[Symbol.iterator](): IterableIterator<number> {
        for (const chunk of this.#chunks) {
            let bits = chunk.bits;
            let offset = 0;
            while (bits !== 0n) {
                if ((bits & 1n) !== 0n) yield chunk.wordIndex * 64 + offset;
                bits >>= 1n;
                offset++;
            }
        }
    }

    /**
     * Whether both bit sets share a node by identity. A representation test used to confirm a no-op
     * avoided copying, not an equality test.
     */
    public sharesStorageWith(other: PersistentChunkedBitSet): boolean {
        return this.#chunks.sharesStructureWith(other.#chunks);
    }

    /**
     * Walk the whole bit set and check its invariants, throwing on the first violation. A defensive
     * audit, not part of normal use.
     */
    public validateStructure(): ChunkedBitSetStatistics {
        let chunkCount = 0;
        let total = 0;
        let previous = -1;
        for (const chunk of this.#chunks) {
            if (chunk.wordIndex <= previous || chunk.bits === 0n || chunk.bits >> 64n !== 0n) {
                throw new Error("The chunked bit set has invalid word ordering or contents.");
            }
            previous = chunk.wordIndex;
            chunkCount++;
            total += popCount(chunk.bits);
        }
        const summary = this.#chunks.measure;
        if (summary.chunkCount !== chunkCount || summary.popCount !== total
            || summary.lastWordIndex !== (chunkCount === 0 ? undefined : previous)) {
            throw new Error("The chunked bit-set annotations disagree with its words.");
        }
        return { chunkCount, popCount: total };
    }

    private locateWord(wordIndex: number): ReturnType<MeasuredSequence<BitSetChunk, BitSetSummary>["locate"]> {
        return this.#chunks.locate((summary) => summary.lastWordIndex !== undefined && summary.lastWordIndex >= wordIndex);
    }

    private combine(other: PersistentChunkedBitSet, operation: BitSetOperation): PersistentChunkedBitSet {
        const left = this.#chunks.toArray();
        const right = other.#chunks.toArray();
        const result: BitSetChunk[] = [];
        let leftIndex = 0;
        let rightIndex = 0;
        while (leftIndex < left.length || rightIndex < right.length) {
            const leftChunk = left[leftIndex];
            const rightChunk = right[rightIndex];
            if (rightChunk === undefined || (leftChunk !== undefined && leftChunk.wordIndex < rightChunk.wordIndex)) {
                if (operation === "union" || operation === "except" || operation === "symmetricExcept") result.push(leftChunk!);
                leftIndex++;
                continue;
            }
            if (leftChunk === undefined || rightChunk.wordIndex < leftChunk.wordIndex) {
                if (operation === "union" || operation === "symmetricExcept") result.push(rightChunk);
                rightIndex++;
                continue;
            }
            const bits = operation === "union" ? leftChunk.bits | rightChunk.bits
                : operation === "intersect" ? leftChunk.bits & rightChunk.bits
                    : operation === "except" ? leftChunk.bits & ~rightChunk.bits
                        : leftChunk.bits ^ rightChunk.bits;
            if (bits !== 0n) result.push({ wordIndex: leftChunk.wordIndex, bits });
            leftIndex++;
            rightIndex++;
        }
        if (sameChunks(result, left)) return this;
        if (result.length === 0) return PersistentChunkedBitSet.empty();
        return new PersistentChunkedBitSet(MeasuredSequence.from(result, bitSetMeasure));
    }
}

/** Immutable root-plus-population-rank cursor over the present bits of a sparse bit set. */
export class PersistentChunkedBitSetCursor {
    public constructor(public readonly set: PersistentChunkedBitSet, public readonly position = 0) {
        if (!Number.isInteger(position) || position < 0 || position > set.count) throw new RangeError("Cursor position is outside the chunked bit set.");
    }
    /** Number of set bits in the bit set version this cursor is positioned in. */
    public get count(): number { return this.set.count; }
    /** Number of set bits. */
    public get size(): number { return this.count; }
    /** Whether the gap precedes the first set bit. */
    public get isAtStart(): boolean { return this.position === 0; }
    /** Whether the gap follows the last set bit. */
    public get isAtEnd(): boolean { return this.position === this.count; }
    /** The set bit immediately before the gap, or `undefined` at the start. */
    public peekPrevious(): { readonly value: number } | undefined { return this.isAtStart ? undefined : { value: this.set.select(this.position - 1) }; }
    /** The set bit immediately after the gap, or `undefined` at the end. */
    public peekNext(): { readonly value: number } | undefined { return this.isAtEnd ? undefined : { value: this.set.select(this.position) }; }
    /**
     * A cursor one position earlier. The receiver is unchanged; movement produces a new cursor over
     * the same version.
     */
    public movePrevious(): PersistentChunkedBitSetCursor { if (this.isAtStart) throw new RangeError("Cursor is already at the start."); return new PersistentChunkedBitSetCursor(this.set, this.position - 1); }
    /** A cursor one position later. The receiver is unchanged. */
    public moveNext(): PersistentChunkedBitSetCursor { if (this.isAtEnd) throw new RangeError("Cursor is already at the end."); return new PersistentChunkedBitSetCursor(this.set, this.position + 1); }
    /** A cursor at the given rank within the same bit set version. */
    public seekRank(position: number): PersistentChunkedBitSetCursor { return position === this.position ? this : new PersistentChunkedBitSetCursor(this.set, position); }
    /** A bit set containing the given set bit; returns the receiver when it is already present. */
    public add(bitIndex: number): PersistentChunkedBitSetCursor { const set = this.set.add(bitIndex); if (set === this.set) return this; const position = bitIndex === 0 ? 0 : this.set.rank(bitIndex - 1); return new PersistentChunkedBitSetCursor(set, position + 1); }
    /** Remove the set bit before the gap and return a cursor in its place. */
    public deletePrevious(): PersistentChunkedBitSetCursor { const bit = this.peekPrevious(); if (bit === undefined) throw new RangeError("No set bit precedes the cursor."); return new PersistentChunkedBitSetCursor(this.set.remove(bit.value), this.position - 1); }
    /** Remove the set bit after the gap and return a cursor in its place. */
    public deleteNext(): PersistentChunkedBitSetCursor { const bit = this.peekNext(); if (bit === undefined) throw new RangeError("No set bit follows the cursor."); return new PersistentChunkedBitSetCursor(this.set.remove(bit.value), this.position); }
    /** The bit set version this cursor is positioned in. */
    public snapshot(): PersistentChunkedBitSet { return this.set; }
}

function requireBitIndex(bitIndex: number): void {
    if (!Number.isInteger(bitIndex) || bitIndex < 0 || bitIndex > maximumBitIndex) {
        throw new RangeError("bitIndex must be a nonnegative signed-32-bit integer.");
    }
}

function popCount(value: bigint): number {
    let remaining = value;
    let count = 0;
    while (remaining !== 0n) { remaining &= remaining - 1n; count++; }
    return count;
}

function sameChunks(left: readonly BitSetChunk[], right: readonly BitSetChunk[]): boolean {
    return left.length === right.length
        && left.every((chunk, index) => chunk.wordIndex === right[index]?.wordIndex && chunk.bits === right[index]?.bits);
}
